include/engine/geospatial_query.hpp

```diff
@@ -47,6 +47,42 @@ template <typename RTreeT, typename DataFacadeT> class GeospatialQuery
         return rtree.SearchInBox(bbox);
     }
 
+    std::vector<PhantomNodeWithDistance>
+    NearestPhantomNodes(const util::Coordinate input_coordinate,
+                        const Approach approach,
+                        const double max_distance,
+                        const boost::optional<Bearing> bearing_with_range,
+                        const boost::optional<bool> use_all_edges) const
+    {
+        auto results = rtree.SearchInRange(
+            input_coordinate,
+            max_distance,
+            [this, approach, &input_coordinate, &bearing_with_range, &use_all_edges, max_distance](
+                const CandidateSegment &segment)
+            {
+                auto invalidDistance =
+                    CheckSegmentDistance(input_coordinate, segment, max_distance);
+                if (invalidDistance)
+                {
+                    return std::pair<bool, bool>{false, false};
+                }
+                auto valid = CheckSegmentExclude(segment) &&
+                             CheckApproach(input_coordinate, segment, approach) &&
+                             (use_all_edges ? HasValidEdge(segment, *use_all_edges)
+                                            : HasValidEdge(segment)) &&
+                             (bearing_with_range ? CheckSegmentBearing(segment, *bearing_with_range)
+                                                 : std::make_pair(true, true));
+                return valid;
+            });
+        auto phantom_nodes = MakePhantomNodes(input_coordinate, results);
+        // there is no guarantee that `SearchInRange` will return results sorted by distance,
+        // but we may rely on it somewhere
+        std::sort(phantom_nodes.begin(),
```

Comment:
I'd be tempted to leave this out, and just see what happens.  We _already_ have some "construction order dependent" results (when there are two paths to a destination of equal weight, the one picked depends on the order the graph was built, not some other stable heuristic), so stability is already not guaranteed.

Reply:
That'd be an interesting insight, indeed.

Reply:
Well, actually after I added it I realised that this method is only used in map matching. For map matching order shouldn't be important I believe. So I think it will be safe to remove it :+1: